cpu/esp32/esp-now/esp_now_netdev.c

```diff
@@ -214,13 +214,16 @@ static IRAM_ATTR void esp_now_recv_cb(const uint8_t *mac, const uint8_t *data, i
     }
 #endif
 
-    if (_esp_now_dev.rx_pkt.len) {
-        /* there is already a packet in receive buffer, we drop the new one */
+    critical_enter();
+    mutex_lock(&_esp_now_dev.dev_lock);
```

Comment:
I'm not sure because we don't have insight to the proprietary `libespnow.a` library, but I would guess that esp_now_recv_cb is called in interrupt context. At least the receiption of a packet is not handled by the wifi-event-loop thread https://github.com/RIOT-OS/RIOT/blob/34b37d18f60e45d07f2bd22b4cdfe6903405ad1e/cpu/esp32/vendor/esp-idf/esp32/event_loop.c#L49 like all other WiFi events. It is also not handled by the `_esp_system_event_handler` function https://github.com/RIOT-OS/RIOT/blob/34b37d18f60e45d07f2bd22b4cdfe6903405ad1e/cpu/esp32/esp-now/esp_now_netdev.c#L264 like the `SYSTEM_EVENT_SCAN_DONE` event which leads to the call of `esp_now_scan_peers_done`.

I suppose that the receiption of an IEEE 802.11 Vendor Action Frame that is used as vehicle for ESP-NOW messages causes just an interrupt in the WiFi hardware which leads to a callback of the registered `esp_now_recv_cb` function by `libespnow.a` in the interrupt context.

The problem is that `mutex_lock` doesn't work at all in the interrupt context. 

If I'm right with my assumption, handling the mutex could be removed here since the only instance that can call `esp_now_recv_cb` is `libespnow.a` triggered by an interrupt from the single WiFi interface. Disabling interrupts should be sufficient to avoid conflicts with subsequent interrupts in that case.

If I'm not right, the order of `critical_enter()` and `mutex_lock` should be changed since `mutex_lock` might lead to thread context switch with disabled interrupts. `mutex_lock` itself prevents inconsistencies by disabling the interrupts during mutex handling.

Reply:
So if I switch the order, it should work in either case? Or would the mutex hurt in the case it's a an ISR?

Reply:
Quite difficult to answer. We have to consider different cases:
- If the mutex is unlocked, `mutex_lock` just marks the mutex as locked and returns and everything is fine.
- If the mutex is locked, `mutex_lock` markes the calling thread as `STATUS_MUTEX_BLOCKED` (in our case the interrupted thread), places it in the waiting list and yields to a thread of higher priority if there is one. However, and here comes the problem, the context switch is not executed immediately when we are in an ISR, but only when the ISR is exited. That is, `mutex_lock` returns with the calling thread marked as `STATUS_MUTEX_BLOCKED` and placed in the waiting list, and the ISR is further executed. At the end of the ISR, the mutex is tried to be unlocked and the first waiting thread is taken from the waiting list. If it is our interrupted thread, it is marked as `STATUS_PENDING` and the context switch at the end of our ISR switches to our interrupted thread and everything is fine.
- But, if there is already a thread in the waiting list with higher priority than the interrupted thread, I'm not sure what happens. Probably, it strongly depends on the priorities of the threads and their order in the waiting queue.

So we have to think about whether the last case can happen at all. We have `mutex_lock` in

- `_recv` and `_send` functions that are called by the `esp_net_if` thread `net_esp_now` with priority 10.
- `esp_now_scan_peers_done` function that is called by the `wifi-event-loop` thread with priority 4.
- `esp_now_recv_cb` function that is probably in interrupt context, where any thread with any priority can be interrupted.

If `wifi-event-loop`, `net_esp_now` is already in waiting queue and a low priority thread is interrupted by incoming packet could be a problem.

There is also a thread 'wifi' with a priority of 1 which is created by the proprietary WiFi libraries using the FreeRTOS adaption layer. Maybe, it is acting as WiFi interrupt handler and calls `esp_now_recv_cb` in thread context. In that case we wouldn't have a problem.

Reply:
But we should change the order of the calls anyway.

Reply:
I added the following to esp_now_recv_cb:

```
if (irq_is_in()) {
    printf(">>>>> esp_now_recv_cb called in isr!\n");
} else {
    printf(">>>>> esp_now_recv_cb called in thread %d!\n", (int)thread_getpid());
}
```

It comes back telling me it's running in thread 8, which is that wifi thread with priority 1.

Order of lock/critical_enter is already fixed.

Reply:
I added the following to esp_now_recv_cb:

```
if (irq_is_in()) {
    printf(">>>>> esp_now_recv_cb called in isr!\n");
} else {
    printf(">>>>> esp_now_recv_cb called in thread %d!\n", (int)thread_getpid());
}
```

It comes back telling me it's running in thread 8, which is that wifi thread with priority 1.

Order of lock/critical_enter is already fixed.